Rewrite HRA

phargogh opened the pull request:
## Description

This PR rewrites HRA, offering a modest (~30%) speedup and (more importantly) making the model easier to maintain.

Fixes #819 
Fixes #751 
Fixes #654 

Outside of the fact that this is a rewrite, notable changes to the source code include:
* A new function aligns the stack of rasters and/or vectors (no rasters strictly required), offering more accurate rasterization as a result.
* A new override function is offered as part of datastack archive construction to allow HRA and other models a way to override the archiving of tables that might refer to other files in nonstandard ways.  See #751 and #342 
* The UI has a new input to allow the user to configure the max number of overlapping stressors.  This affects the breakpoints in the high/med/low risk classifications and is related to #819 

I am working on improvements to the user's guide to clarify a few things and will update the UG revision when they are complete.

## Checklist
- [x] Updated HISTORY.rst (if these changes are user-facing)
- [ ] Updated the user's guide (if needed)
- [x] Tested the affected models' UIs (if relevant)

davemfish reviewed (changes requested):
Hey @phargogh , thanks for taking on this rewrite! This was a pleasure to read. The program is very well-organized and should be very easy to maintain. I have a bunch of minor comments throughout, and a couple of general issues:

1. We'll need an update to `workbench/src/renderer/ui_config.js` for the new `n_overlapping_stressors` arg. That should be why the workbench tests are failing now.
2. I tried running the sample data with `n_workers: 2` and encountered this error:
```
Traceback (most recent call last):
  File "/home/dmf/projects/invest/env/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/dmf/projects/invest/env/lib/python3.8/site-packages/natcap/invest/hra.py", line 1128, in _create_summary_statistics_file
    valid_pixels = _read_block(raster, block_info,
  File "/home/dmf/projects/invest/env/lib/python3.8/site-packages/natcap/invest/hra.py", line 1051, in _read_block
    band = raster.GetRasterBand(1)
AttributeError: 'NoneType' object has no attribute 'GetRasterBand'
```
Can you reproduce this?

phargogh commented on src/natcap/invest/datastack.py:
```diff
@@ -223,6 +218,26 @@ def build_datastack_archive(args, model_name, datastack_path):
+        # for the other args keys.
+        override_funcname = f'_override_datastack_archive_{key}'
+        if hasattr(module, override_funcname):
+            LOGGER.info(f'Using model override function for key {key}')
```
I agree, the use of `INFO` priority was a little mindless on my part. b1e3cb907

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+                }
+            },
+            "about": gettext(
+                "A GDAL-supported vector file containing feature containing "
```
692419dbe

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+                "The number of overlapping stressors to consider as "
+                "'maximum' when reclassifying risk scores into "
+                "high/medium/low.  Affects the breaks between risk "
+                "classifications."),
```
Yep, I agree.
- [ ] Add guidance for how to choose this value in the UG

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+            be used as a way of identifying each individual shape.
+        args['n_overlapping_stressors'] (number): If provided, this
+            number will be used in risk reclassification instead of the
+            calculated maximum number of stressor layers that overlap.
```
Yep!  It used to be optional in an earlier dev version and has since been made required. c8306917d

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+            contains the set of criteria ranking of each stressor on each
+            habitat.
+        args['resolution'] (int): a number representing the desired pixel
+            #dimensions of output rasters in meters.
```
Whoops 53f47d8a6

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+        max_pairwise_risk = max_rating * max_rating
+    elif args['risk_eq'].lower() == 'euclidean':
+        max_pairwise_risk = math.sqrt(
+            ((max_rating - 1) ** 2) + ((max_rating - 1) ** 2))
```
Oooo yes, absolutely.  The UG is not very clear on what this score is, so logging it would absolutely help see what the model's doing.  eece5c048

phargogh commented on Makefile:
```diff
@@ -10,7 +10,7 @@ GIT_TEST_DATA_REPO_REV      := ac7023d684478485fea89c68f8f4154163541e1d
 GIT_UG_REPO                 := https://github.com/natcap/invest.users-guide
 GIT_UG_REPO_PATH            := doc/users-guide
-GIT_UG_REPO_REV             := eef0643feea534db130198e81a69a0590c1a8480
+GIT_UG_REPO_REV             := eeeb7c34d5b35a9e6cbbdac1c6404522c9b5f9a0
```
Yes!  Very happy to report that no test data are needed for the model.

Sample data repo has been updated in ad1ae46ff, test data removed in c4d2bfb76

New PRs on bitbucket for these:
* https://bitbucket.org/natcap/invest-test-data/pull-requests/26/remove-hra-subdirectory
* https://bitbucket.org/natcap/invest-sample-data/pull-requests/26/add-hra-n_overlapping_stressors-parameter

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+    graph = taskgraph.TaskGraph(taskgraph_working_dir, n_workers)
+
+    # parse the info table and get habitats, stressors
+    habitats, stressors = _parse_info_table(args['info_table_path'])
```
Yeah, that's a good point.  How about a rename to `habitats_info` and `stressors_info`, to clarify that these are not just sequences of strings?

In case that's a good solution, implemented in 85229f68a

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+            parseable (for a program) format.
+
+    Returns:
+        ``None``, but the table written out to ``target_composite_csv_path`` is
```
Whoops, you're right!  c77530e1c

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+        args['criteria_table_path'], composite_criteria_table_path)
+
+    # Validate that habitats and stressors match precisely.
+    for label, info_set, criteria_set in [
```
👍 1d8ee14cc

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+        name = f'{habitat}-{stressor}-{criterion}'
+        spatial_criteria_attrs[name] = {
+            'name': name,
+            'path': rating,  # Previously validated to be a GIS type.
```
Yeah, this was not implemented.  I added validation for this in `_parse_criteria_table` in c138180bb

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+                    kwargs={
+                        'source_raster_path': source_filepath,
+                        'target_filepath': rewritten_raster_path,
+                    },
```
> All python function positional args are actually treated as named args if called with an unpacked dict?

Yes!  Any named parameter to a python function can be passed by name.

There are a few caveats of course, like kwargs have to come after positional args, and certain built-in functions (like `range`, I think) do not accept kwargs at all.

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+
+                # This rather complicated filter just grabs the rows matching
+                # this habitat, stressor and criteria type.  It's the pandas
+                # equivalent of SELECT * FROM criteria_df WHERE the habitat,
```
You know, reworking HRA here gave me a new appreciation for them.  Boolean filtering on a table like we have here is one of the things that is really pretty elegant.  Nothing against SQL, but I'd WAY rather do this filtering on a plain-text CSV!

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+                    'local_op': _reclassify_score,
+                    'target_raster_path': reclassified_pairwise_risk_path,
+                    'datatype_target': _TARGET_GDAL_TYPE_BYTE,
+                    'nodata_target': _TARGET_NODATA_BYTE
```
I'm glad you like it!

I also found it helpful to write things out like this, but I also have 2 minor gripes:
* It makes `execute` about twice as long
* It's tedious to write out all the parameter names (and I really wish my editor would do it for me)

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+
+    pygeoprocessing.raster_calculator(
+        [(path, 1) for path in raster_path_list],
+        _sum_op, target_result_path, gdal.GDT_Float32, target_nodata)
```
Yeah, I agree.  It's weird to separate them.  I added the target datatype as a parameter to the function in f5824d022

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+            'datatype_target': _TARGET_GDAL_TYPE_BYTE,
+            'nodata_target': _TARGET_NODATA_BYTE,
+        },
+        task_name=f'Reclassify risk for {habitat}/{stressor}',
```
Yep!  911dc4ee2

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+    out_dataframe.to_csv(target_summary_csv_path, index=False)
+
+
+def _rasterize_aoi_regions(source_aoi_vector_path, habitat_mask_raster,
```
Good point!  986b27a31

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+        source_aoi_vector_path (string): The path to the source AOI vector
+            containing AOI geometries and (optionally) a "NAME" column
+            (case-insensitive).
+        habitat_mask_path (string): The path to a raster where pixel values of
```
Yep, this is also changed in 986b27a31

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+    layer.StartTransaction()
+    # 0 represents field index 0, into which pixel values will be written.
+    gdal.Polygonize(band, mask_band, layer, 0)
+    layer.CommitTransaction()
```
You know, I'm not sure if we ever settled on a strict expectation of this for InVEST, so I added the ` * = None`s to 3387da788.

Theoretically, these objects should be cleaned up due to reference counting when they go out of scope, but there's no harm in setting them to `None` anyways.

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+
+    Functionally, this is a modified version of
+    ``pygeoprocessing.zonal_statistics`` that doesn't force re-rasterization of
+    subregions.
```
Yeah, this functionality follows what's in the current HRA (which I'm not committed to), which attempts to avoid potentially many re-rasterizations in order to do `zonal_statistics`.  The only differentiation between this and `pygeoprocessing.zonal_statistics` is that this function attempts to support grouping features with the same name.

That being said, all of this rasterization/custom stats workaround is a pain and probably one of the harder things to maintain in this model and it doesn't have to be.  I'll rework this into `zonal_statistics` and get back to you.

phargogh replied:
@davemfish after digging into this a bit more, the core issue is that `zonal_statistics` does not currently provide enough information to be able to build the `R_%LOW`, `R_%MED`, and `R_%HIGH` columns; they cannot be generally derived from the `sum` and `count` fields returned from zonal_stats.  I've put together a [gist going into a bit more detail about it here](https://gist.github.com/phargogh/aff32bb689220fb317b0a5196c7cf307).

So if you're OK with it, I'd prefer to leave this function as-is until I can propose a solution within `zonal_statistics` and we can ship it.

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+
+                    # Old HRA included 0-value pixels in  the R_%_LOW bucket so
+                    # we do here as well.
+                    stats['R_N_LOW'] += numpy.count_nonzero(classes_block <= 1)
```
Yeah, this looks contradictory but it should work as expected.  We're calling `numpy.count_nonzero` on a boolean index, where the boolean index is 1 where `classes_block <= 1`.  So therefore pixels in `classes_block` with a pixel value of 0 would be tallied as `True` and therefore `1`.

To avoid this confusion, I've updated this to use `numpy.sum` in 38839d1c9

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+        disjoint_layer.CommitTransaction()
+        gdal.RasterizeLayer(
+            fid_raster, [1], disjoint_layer,
+            options=["ALL_TOUCHED=FALSE", "ATTRIBUTE=FID"])
```
Whoops, yes, I think this makes sense to use `ALL_TOUCHED=FALSE` in both these cases for consistency here _and also_ for consistency with `zonal_statistics`.  Patched in 214ec93b3

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+    """
+    # Step 1: Create a bounding box of the union of all input spatial rasters
+    # and vectors.  To be safe, we're assuming that the source SRS of a dataset
+    # might be different from the target SRS and this is taken into account.
```
Makes sense to me, and thanks!  patched in n

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+        raster_info = pygeoprocessing.get_raster_info(source_raster_path)
+
+        # Integer (discrete) rasters should be nearest-neighbor, continuous
+        # rasters should be interpolated with bilinear.
```
Whoops!  Yes, patched in 157ad7f97

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+        else:
+            # If the simplification didn't work for whatever reason, fall back
+            # to the original geometry.
+            LOGGER.warning(
```
I really don't know how common this sort of thing might be.  It's behavior that's included in the current edition of HRA and seems reasonable to include here.  In my experience with OGR geometries, the GEOS built here is somehow less tolerant than the one that Shapely provides, so I'm inclined to err on the side of caution.

The consequences of a failed simplification are losing a runtime speedup, which I agree is more of a `DEBUG` state than a `WARNING`.  I've updated this logging level to `DEBUG` in 80ca85719

phargogh commented on src/natcap/invest/hra.py:
```diff
@@ -1,3253 +1,2424 @@
+    Returns:
+        ``None``.
+    """
+    if preserve_columns is None:
```
I haven't observed it taking a long time, but it's always possible that it could if the user provides a complicated geometry.  Adding a logging statement in f1074330b

phargogh commented:
OK! @davemfish  I think I've addressed all of your comments attached to specific lines.  Additionally:

* 37cfd5116 contains the workbench fix you mentioned (thanks for that!)
* 43d957761 fixes the taskgraph error you mentioned, which was caused (as usual) by some tasks being a dependency but not being tracked as such in the graph.

I also still have a [note-to-self about modifying the UG to include some text about the new `n_overlapping_stressors`](https://github.com/natcap/invest/pull/1026#discussion_r929256348) parameter and am waiting to hear back from Jade about this.
- [ ] update the UG about `n_overlapping_stressors`

Also, I had previously been thinking about this rewrite as basically a very large bugfix and so the PR target is set to `main`.  Does that still feel right?  Just wanted to be sure to mention it in case it didn't.

davemfish commented:
> OK! @davemfish I think I've addressed all of your comments attached to specific lines. The `zonal_statistics` thread is particularly interesting and I would definitely appreciate another set of eyes on my conclusion to make sure I'm not missing anything.
> 
> Re: the other couple things you mentioned that aren't tied to a particular line:
> 
> * [37cfd51](https://github.com/natcap/invest/commit/37cfd51166c9d56d040389343ce1ed439788af55) contains the workbench fix you mentioned (thanks for that!)
> * [43d9577](https://github.com/natcap/invest/commit/43d957761de07d29f836c8a86925ead0210f774a) fixes the taskgraph error you mentioned, which was caused (as usual) by some tasks being a dependency but not being tracked as such in the graph.
> * I had an issue in the updated datastack, which has been fixed in this PR: https://bitbucket.org/natcap/invest-sample-data/pull-requests/27
> 
> I also still have a [note-to-self about modifying the UG to include some text about the new `n_overlapping_stressors`](https://github.com/natcap/invest/pull/1026#discussion_r929256348) parameter and am waiting to hear back from Jade about this.
> 
> * [ ]  update the UG about `n_overlapping_stressors`
> 
> Also, I had previously been thinking about this rewrite as basically a very large bugfix and so the PR target is set to `main`. Does that still feel right? Just wanted to be sure to mention it in case it didn't.

Thanks @phargogh your explanation about the zonal stats makes total sense, thanks for that write-up!

As for target being `main`, that's a good question. I think the only thing that gives me pause is the new `n_overlapping_stressors` parameter, since that's an API change. But, I also think it's unlikely that we do a bugfix release from `main` before we do `3.12`, so these changes probably end up in `3.12` no matter what the target branch is. We can chat about it real quick on the call today and decide.

davemfish reviewed (approved):
It all checks out!
